### PR TITLE
rustc_codegen_ssa: generate MSVC cleanup pads on demand, like GNU landing pads.

### PR DESCRIPTION
**Pull request by eddyb**
This unblocks #84993 in terms of codegen tests, as it brings the MSVC-style (`cleanup_pad`) EH (LLVM) block order in line with the GNU-style (`landing_pad`) EH (LLVM) block order, by having both of them be on-demand (instead of MSVC-style being eager and GNU-style lazy/on-demand).

It also unifies the two implementations a bit, similar to #84699, but in the opposite direction (as that attempt made both kinds of EH pads eagerly built).

~~Opening as draft because I haven't done enough Windows testing just yet, of both this PR, and of #84993 rebased on it.~~ (**EDIT**: seems to be working as expected)

r? @nagisa

### CODE REVIEW

**Comment by eddyb** [comment]
On MSVC, I was able to bootstrap, and build `cargo`, `ripgrep` and `syn`, with this PR.

Also, rebasing #84993 on this PR did indeed fix the test failure there (i.e. MSVC and GNU styles indeed agree on the LLVM block order).

So I'm un-drafting this PR, it's ready for review I think.

Also, in case anything goes wrong, in the interest of easy bisecting:
@bors rollup=never

File: compiler/rustc_codegen_ssa/src/mir/mod.rs [hunk 1]
```diff
@@ -49,12 +46,13 @@ pub struct FunctionCx<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> {
...
-    /// BB. This is initialized as we compute the funclets' head block in RPO.
+    /// When targeting MSVC, this stores the cleanup info for each funclet BB.
+    /// This is initialized at the same time as the `landing_pads` entry for the
+    /// funclets' head block, i.e. when needed by an unwind / `cleanup_ret` edge.
     funclets: IndexVec<mir::BasicBlock, Option<Bx::Funclet>>,
 
-    /// This stores the landing-pad block for a given BB, computed lazily on GNU
-    /// and eagerly on MSVC.
+    /// This stores the cached landing/cleanup pad block for a given BB.
+    // FIXME(eddyb) rename this to `eh_pads`.
```
Comment (nagisa) [inline compiler/rustc_codegen_ssa/src/mir/mod.rs:55]: Any reason these renames aren't happening now?

Comment (eddyb) [reply compiler/rustc_codegen_ssa/src/mir/mod.rs:55]: I can add a commit for them if you want, no particular reason, was focusing on getting it working and tested first.

**Pull request merged**